metrics-facade-base/src/test/java/com/ringcentral/platform/metrics/instrument/executors/MonitoredScheduledExecutorServiceTest.java

```diff
@@ -0,0 +1,379 @@
+        // when
+        timeNanosProvider.increaseSec(1);
+
+        // then
+        verify(executionTimer).stopwatch(LABEL_VALUES);
+        verify(runningCounter).inc(LABEL_VALUES);
+        verify(runningCounter).dec(LABEL_VALUES);
+        verify(completedRate).mark(LABEL_VALUES);
+        verify(executionStopwatch).stop();
+        verify(periodPercentHistogram).update(100L, LABEL_VALUES);
+        // verify(scheduledOverrunRate).mark(LABEL_VALUES);
+        verifyNoMoreMetricInteractions();
+    }
+
+    @Test
+    public void scheduleAtFixedRate_Overrun() {
+        // given
+        verifyNoMoreMetricInteractions();
+        when(executionStopwatch.stop()).thenReturn(SECONDS.toNanos(15L));
+        Runnable task = mock(Runnable.class);
+
+        // when
+        monitoredExecutor.scheduleAtFixedRate(task, 10, 10, SECONDS);
+
+        // then
+        verifyNoMoreMetricInteractions();
+
+        // when
+        timeNanosProvider.increaseSec(9);
+
+        // then
+        verifyNoMoreMetricInteractions();
+
+        // when
+        timeNanosProvider.increaseSec(1);
+
+        // then
+        verify(executionTimer).stopwatch(LABEL_VALUES);
+        verify(runningCounter).inc(LABEL_VALUES);
+        verify(runningCounter).dec(LABEL_VALUES);
+        verify(completedRate).mark(LABEL_VALUES);
+        verify(executionStopwatch).stop();
+        verify(periodPercentHistogram).update(150L, LABEL_VALUES);
+        verify(scheduledOverrunRate).mark(LABEL_VALUES);
+
+        verifyNoMoreMetricInteractions();
+    }
+
+    @Test
+    public void invokeAll() throws Exception {
+        // given
+        verifyNoMoreMetricInteractions();
+
+        Callable<Integer> task_1 = mock(Callable.class);
+        when(task_1.call()).thenReturn(1);
+
+        Callable<Integer> task_2 = mock(Callable.class);
+        when(task_2.call()).thenReturn(2);
+
+        // when
+        assertThat(monitoredExecutor.invokeAll(List.of(task_1, task_2)).stream().map(MonitoredScheduledExecutorServiceTest::value).collect(toList()), is(List.of(1, 2)));
+
+        // then
+        verifyNoMoreMetricInteractions();
+    }
+
+    private void verifyNoMoreMetricInteractions() {
+        verifyNoMoreInteractions(submittedRate, runningCounter, completedRate, idleTimer, idleStopwatch, executionTimer, executionStopwatch);
+        verifyNoMoreInteractions(scheduledOnceRate, scheduledRepetitivelyRate, scheduledOverrunRate, periodPercentHistogram);
+    }
+
+    private static <T> CompletableFuture<T> completedFutureFor(Callable<T> task) {
+        try {
+            return completedFuture(task.call());
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Integer value(Future<Integer> future) {
+        try {
+            return future.get();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    class ScheduledExecutorServiceImpl extends TestScheduledExecutorService {
```

Comment:
Could we somehow make it available in **_metrics-facade-test-support_** module to avoid duplication?

Reply:
I haven't been able to come up with an appropriate solution. :) There is actually a full copy of this class in the metrics-facade-test-support module, but the problem is that the metrics-facade-test-support module requires metrics-facade-base as a dependency, creating a cyclic dependency issue. If you know how this can be resolved, please let me know, and I'd be happy to fix it...